### PR TITLE
Fixed #35099 -- Prevented mutating queryset when combining with & and | operators.

### PR DESCRIPTION
**Pull request by Hisham-Pak**
Thanks to Alan for reporting this issue.

Ticket: https://code.djangoproject.com/ticket/35099

### CODE REVIEW

**Comment by Hisham-Pak** [comment]
Unrelated CI failure

File: tests/queries/tests.py [hunk 1]
```diff
@@ -1357,6 +1357,26 @@ def test_negate_field(self):
...
+
+        all_authors | authors_with_extrainfo
+        all_authors & authors_with_extrainfo
+
+        authors_without_extrainfo = all_authors.exclude(pk__in=authors_with_extrainfo)
+        items_after = Item.objects.filter(creator__in=authors_without_extrainfo)
+        query_after = items_after.query
+
+        self.assertCountEqual(items_before, items_after)
+        self.assertEqual(str(query_before), str(query_after))
```
Comment (felixxm) [inline tests/queries/tests.py:1378]: There is no need to compare SQL queries.

**Comment by Hisham-Pak** [comment]
@felixxm Thanks for the review. I think  `assertQuerySetEqual` would've passed too.

**Comment by Hisham-Pak** [comment]
I remember postgres and oracle failures were because these queries behaved differently on different backends. So shouldn't these asserts be replaced with more dynamic checks?

**Comment by felixxm** [comment]
> I remember postgres and oracle failures were because these queries behaved differently on different backends. So shouldn't these asserts be replaced with more dynamic checks?

This has nothing to do with backend. Queryset in tests doesn't make much sense:
```
        authors_with_extrainfo = Author.objects.filter(
            Exists(ExtraInfo.objects.filter(id=OuterRef("id")))
        )
```
because you compare IDs of different objects. I will correct it.

**Pull request merged**